File: java/socketd/src/main/java/org/noear/socketd/transport/core/internal/ProcessorDefault.java
package org.noear.socketd.transport.core.internal;

import org.noear.socketd.exception.SocketdAlarmException;
import org.noear.socketd.exception.SocketdConnectionException;
import org.noear.socketd.transport.core.*;
import org.noear.socketd.transport.core.listener.SimpleListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

/**
 * 协议处理器默认实现
 *
 * @author noear
 * @since 2.0
 */
public class ProcessorDefault implements Processor {
    private static Logger log = LoggerFactory.getLogger(ProcessorDefault.class);

    private Listener listener = new SimpleListener();

    /**
     * 设置监听
     */
    @Override
    public void setListener(Listener listener) {
        if (listener != null) {
            this.listener = listener;
        }
    }

    /**
     * 接收处理
     */
    public void onReceive(ChannelInternal channel, Frame frame)  {
        if (log.isDebugEnabled()) {
            if (channel.getConfig().clientMode()) {
                log.debug("C-REV:{}", frame);
            } else {
                log.debug("S-REV:{}", frame);
            }
        }

        if (frame.flag() == Flags.Connect) {
            //if server
            HandshakeDefault handshake = new HandshakeDefault(frame.message());
            channel.setHandshake(handshake);

            //开始打开（可用于 url 签权）//禁止发消息
            channel.onOpenFuture((r,e)-> {
                if (e == null) {
                    //如果无异常
                    if (channel.isValid()) {
                        //如果还有效，则发送链接确认
                        try {
                            channel.sendConnack(frame.message()); //->Connack
                        } catch (Throwable err) {
                            onError(channel, err);
                        }
                    }
                } else {
                    //如果有异常
                    if (channel.isValid()) {
                        //如果还有效，则关闭通道
                        channel.close(Constants.CLOSE3_ERROR);
                        onCloseInternal(channel);
                    }
                }
            });
            onOpen(channel);
        } else if (frame.flag() == Flags.Connack) {
            //if client
            HandshakeDefault handshake = new HandshakeDefault(frame.message());
            channel.setHandshake(handshake);

            onOpen(channel);
        } else {
            if (channel.getHandshake() == null) {
                channel.close(Constants.CLOSE1_PROTOCOL);

                if(frame.flag() == Flags.Close){
                    //说明握手失败了
                    throw new SocketdConnectionException("Connection request was rejected");
                }

                if (log.isWarnEnabled()) {
                    log.warn("{} channel handshake is null, sessionId={}",
                            channel.getConfig().getRoleName(),
                            channel.getSession().sessionId());
                }
                return;
            }

            try {
                switch (frame.flag()) {
                    case Flags.Ping: {
                        channel.sendPong();
                        break;
                    }
                    case Flags.Pong: {
                        break;
                    }
                    case Flags.Close: {
                        //关闭通道
                        channel.close(Constants.CLOSE1_PROTOCOL);
                        onCloseInternal(channel);
                        break;
                    }
                    case Flags.Alarm: {
                        //结束流，并异常通知
                        SocketdAlarmException exception = new SocketdAlarmException(frame.message());
                        StreamInternal stream = channel.getConfig().getStreamManger().getStream(frame.message().sid());
                        if (stream == null) {
                            onError(channel, exception);
                        } else {
                            channel.getConfig().getStreamManger().removeStream(frame.message().sid());
                            stream.onError(exception);
                        }
                        break;
                    }
                    case Flags.Message:
                    case Flags.Request:
                    case Flags.Subscribe: {
                        onReceiveDo(channel, frame, false);
                        break;
                    }
                    case Flags.Reply:
                    case Flags.ReplyEnd: {
                        onReceiveDo(channel, frame, true);
                        break;
                    }
                    default: {
                        channel.close(Constants.CLOSE2_PROTOCOL_ILLEGAL);
                        onCloseInternal(channel);
                    }
                }
            } catch (Throwable e) {
                onError(channel, e);
            }
        }
    }

    private void onReceiveDo(ChannelInternal channel, Frame frame, boolean isReply) throws IOException {
        //如果启用了聚合!
        if(channel.getConfig().getFragmentHandler().aggrEnable()) {
            //尝试聚合分片处理
            String fragmentIdxStr = frame.message().meta(EntityMetas.META_DATA_FRAGMENT_IDX);
            if (fragmentIdxStr != null) {
                //解析分片索引
                int index = Integer.parseInt(fragmentIdxStr);
                Frame frameNew = channel.getConfig().getFragmentHandler().aggrFragment(channel, index, frame.message());

                if (frameNew == null) {
                    return;
                } else {
                    frame = frameNew;
                }
            }
        }

        //执行接收处理
        if (isReply) {
            channel.retrieve(frame);
        } else {
            onMessage(channel, frame.message());
        }
    }


    /**
     * 打开时
     *
     * @param channel 通道
     */
    @Override
    public void onOpen(ChannelInternal channel) {
        channel.getConfig().getChannelExecutor().submit(() -> {
            try {
                listener.onOpen(channel.getSession());
                channel.doOpenFuture(true, null);
            } catch (Throwable e) {
                if (log.isWarnEnabled()) {
                    log.warn("{} channel listener onOpen error",
                            channel.getConfig().getRoleName(), e);
                }
                channel.doOpenFuture(false, e);
            }
        });
    }

    /**
     * 收到消息时
     *
     * @param channel 通道
     * @param message 消息
     */
    @Override
    public void onMessage(ChannelInternal channel, Message message) {
        channel.getConfig().getChannelExecutor().submit(() -> {
            try {
                listener.onMessage(channel.getSession(), message);
            } catch (Throwable e) {
                if (log.isWarnEnabled()) {
                    log.warn("{} channel listener onMessage error",
                            channel.getConfig().getRoleName(), e);
                }
                onError(channel, e);
            }
        });
    }

    /**
     * 关闭时
     *
     * @param channel 通道
     */
    @Override
    public void onClose(ChannelInternal channel) {
        if (channel.isClosed() == 0) {
            onCloseInternal(channel);
        }
    }

    /**
     * 关闭时（内部处理）
     *
     * @param channel 通道
     */
    private void onCloseInternal(ChannelInternal channel){
        listener.onClose(channel.getSession());
    }

    /**
     * 出错时
     *
     * @param channel 通道
     * @param error   错误信息
     */
    @Override
    public void onError(ChannelInternal channel, Throwable error) {
        listener.onError(channel.getSession(), error);
    }
}

File: python/socketd/socketd/transport/core/internal/ProcessorDefault.py
import asyncio
from abc import ABC

from loguru import logger

from socketd.transport.core.Channel import Channel
from socketd.transport.core.Handshake import Handshake
from socketd.transport.core.Processor import Processor
from socketd.transport.core.Costants import Flag
from socketd.transport.core.SimpleListener import SimpleListener
from socketd.transport.core.entity.Entity import EntityMetas
from socketd.transport.core.entity.Frame import Frame


class ProcessorDefault(Processor, ABC):

    def __init__(self):
        self.listener = SimpleListener()
        self.log = logger.opt()

    def set_listener(self, listener):
        if listener is not None:
            self.listener = listener

    async def on_receive(self, channel: Channel, frame):
        if frame.get_flag() == Flag.Connect:
            connectMessage = frame.get_message()
            channel.set_handshake(Handshake(connectMessage))
            await channel.send_connack(connectMessage)
            await self.on_open(channel.get_session())
        elif frame.get_flag() == Flag.Connack:
            message = frame.get_message()
            channel.set_handshake(Handshake(message))
            await self.on_open(channel.get_session())
        else:
            if channel.get_handshake() is None:
                await channel.close()
                self.log.warning("Channel handshake is None, sessionId={}", channel.get_session().get_session_id())
                return

            channel.set_live_time()

            try:
                if frame.get_flag() == Flag.Ping:
                    await channel.send_pong()
                elif frame.get_flag() == Flag.Pong:
                    pass
                elif frame.get_flag() == Flag.Close:
                    await channel.close()
                    self.on_close(channel.get_session())
                elif frame.get_flag() in [Flag.Message, Flag.Request, Flag.Subscribe]:
                    await self.on_receive_do(channel, frame, False)
                elif frame.get_flag() in [Flag.Reply, Flag.ReplyEnd]:
                    await self.on_receive_do(channel, frame, True)
                else:
                    await channel.close()
                    self.on_close(channel.get_session())
            except Exception as e:
                logger.error(e)
                self.on_error(channel.get_session(), e)

    async def on_receive_do(self, channel: Channel, frame: Frame, isReply):
        fragmentIdxStr = frame.get_message().get_entity().get_meta(EntityMetas.META_DATA_FRAGMENT_IDX)
        if fragmentIdxStr is not None:
            index = int(fragmentIdxStr)
            frameNew: Frame = channel.get_config().get_fragment_handler().aggrFragment(channel, index, frame.get_message())
            if frameNew is None:
                return
            else:
                frame = frameNew

        if isReply:
            await channel.retrieve(frame, lambda error: self.on_error(channel, error))
        else:
            await self.on_message(channel, frame.get_message())

    async def on_open(self, session):
        await self.listener.on_open(session)

    async def on_message(self, channel: Channel, message):
        await asyncio.get_running_loop().run_in_executor(channel.get_config().get_executor(),
                                                   lambda: asyncio.run(self.listener.on_message(
                                                       channel.get_session(), message)))

    def on_close(self, session):
        self.listener.on_close(session)

    def on_error(self, session, error):
        self.listener.on_error(session, error)
